server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,235 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
+		return nil, err
+	}
+	d, err := digest.Compute(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	if err := ul.Upload(d, NewBytesReadSeekCloser(data)); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (ul *BatchCASUploader) UploadFile(path string) (*repb.Digest, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	d, err := digest.Compute(f)
+	if err != nil {
+		return nil, err
+	}
+	// Note: uploader.Upload will close the file.
+	if err := ul.Upload(d, f); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+type bytesReadSeekCloser struct {
+	io.ReadSeeker
+}
+
+func NewBytesReadSeekCloser(b []byte) io.ReadSeekCloser {
+	return &bytesReadSeekCloser{bytes.NewReader(b)}
+}
+func (*bytesReadSeekCloser) Close() error { return nil }
+
+func (ul *BatchCASUploader) flushCurrentBatch() {
+	req := ul.req
+	ul.req = &repb.BatchUpdateBlobsRequest{InstanceName: ul.instanceName}
+	ul.currentBatchSize = 0
+	ul.eg.Go(func() error {
+		rsp, err := ul.casClient.BatchUpdateBlobs(ul.ctx, req)
+		if err != nil {
+			return err
+		}
+		for _, fileResponse := range rsp.GetResponses() {
+			if fileResponse.GetStatus().GetCode() != int32(codes.OK) {
+				return gstatus.Error(codes.Code(fileResponse.GetStatus().GetCode()), fmt.Sprintf("Error uploading file: %v", fileResponse.GetDigest()))
+			}
+		}
+		return nil
+	})
+}
+
+func (ul *BatchCASUploader) Wait() error {
+	if len(ul.req.GetRequests()) > 0 {
+		ul.flushCurrentBatch()
+	}
+	return ul.eg.Wait()
+}
+
+// UploadDirectoryToCAS uploads all the files in a given directory to the CAS
+// as well as the directory structure, and returns the digest of the root
+// Directory proto that can be used to fetch the uploaded contents.
+func UploadDirectoryToCAS(ctx context.Context, env environment.Env, instanceName, rootDirPath string) (*repb.Digest, error) {
+	ul, err := NewBatchCASUploader(ctx, env, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Recursively find and upload all descendant dirs.
+	visited, rootDirectoryDigest, err := uploadDir(ul, rootDirPath, nil /*=visited*/)
+	if err != nil {
+		return nil, err
+	}
+	if len(visited) == 0 {
+		return nil, status.InternalError("empty directory list after uploading directory tree; this should never happen")
+	}
+	// Upload the tree, which consists of the root dir as well as all descendant
+	// dirs.
+	rootTree := &repb.Tree{Root: visited[0], Children: visited[1:]}
+	if _, err := ul.UploadProto(rootTree); err != nil {
+		return nil, err
+	}
+	if err := ul.Wait(); err != nil {
+		return nil, err
+	}
+	return rootDirectoryDigest, nil
+}
+
+func uploadDir(ul *BatchCASUploader, dirPath string, visited []*repb.Directory) ([]*repb.Directory, *repb.Digest, error) {
+	dir := &repb.Directory{}
+	// Append the directory before doing any other work, so that the root
+	// directory is located at visited[0] at the end of recursion.
+	visited = append(visited, dir)
+	infos, err := ioutil.ReadDir(dirPath)
```

Comment:
The Go team is working on deprecating ioutil [1] so it'd be nice to use the "newer" thing (os.ReadDir) from the beginning: 

[1] https://github.com/golang/go/issues/42026

Reply:
Done